core/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java

```diff
@@ -396,60 +413,99 @@ static boolean shouldExecuteReplicaItem(final BulkItemRequest request, final int
         }
         return replica.index(operation);
     }
 
+    /** Utility method to prepare an index operation on replica shards */
+    static Engine.Index prepareIndexOperationOnReplica(
+            DocWriteResponse primaryResponse,
+            IndexRequest request,
+            IndexShard replica) {
+
+        final ShardId shardId = replica.shardId();
+        final long version = primaryResponse.getVersion();
+        final long seqNo = primaryResponse.getSeqNo();
+        final SourceToParse sourceToParse =
+                SourceToParse.source(SourceToParse.Origin.REPLICA, shardId.getIndexName(),
+                        request.type(), request.id(), request.source(), request.getContentType())
+                .routing(request.routing()).parent(request.parent());
+        final VersionType versionType = request.versionType().versionTypeForReplicationAndRecovery();
+        assert versionType.validateVersionForWrites(version);
+
+        return replica.prepareIndexOnReplica(sourceToParse, seqNo, version, versionType,
+                request.getAutoGeneratedTimestamp(), request.isRetry());
+    }
+
     /** Utility method to prepare an index operation on primary shards */
     static Engine.Index prepareIndexOperationOnPrimary(IndexRequest request, IndexShard primary) {
-        SourceToParse sourceToParse =
-            SourceToParse.source(SourceToParse.Origin.PRIMARY, request.index(), request.type(), request.id(), request.source(),
-                request.getContentType()).routing(request.routing()).parent(request.parent());
-        return primary.prepareIndexOnPrimary(sourceToParse, request.version(), request.versionType(), request.getAutoGeneratedTimestamp(), request.isRetry());
+        final SourceToParse sourceToParse =
+                SourceToParse.source(SourceToParse.Origin.PRIMARY, request.index(), request.type(),
+                        request.id(), request.source(), request.getContentType())
+                .routing(request.routing()).parent(request.parent());
+        return primary.prepareIndexOnPrimary(sourceToParse, request.version(), request.versionType(),
+                request.getAutoGeneratedTimestamp(), request.isRetry());
     }
 
     /** Executes index operation on primary shard after updates mapping if dynamic mappings are found */
     public static Engine.IndexResult executeIndexRequestOnPrimary(IndexRequest request, IndexShard primary,
                                                                   MappingUpdatePerformer mappingUpdater) throws Exception {
-        MappingUpdatePerformer.MappingUpdateResult result = mappingUpdater.updateMappingsIfNeeded(primary, request);
-        if (result.isFailed()) {
-            return new Engine.IndexResult(result.failure, request.version());
+        // Update the mappings if parsing the documents includes new dynamic updates
+        try {
+            mappingUpdater.updateMappingsIfNeeded(primary, request);
```

Comment:
see [comment](https://github.com/elastic/elasticsearch/pull/23665/files#r108282321). IMO we should not be calling `prepareIndexOperationOnPrimary` in `ConcreteMappingUpdatePerformer#updateMappingsIfNeeded` and `ConcreteMappingUpdatePerformer#verifyMappings` instead `prepareIndexOperationOnPrimary` should be called once here and then the resulting `operation` should be passed down to `ConcreteMappingUpdatePerformer`